src/core/net/netif.cpp

```diff
@@ -120,28 +169,83 @@ bool Netif::IsMulticastSubscribed(const Address &aAddress) const
 {
     bool rval = false;
 
-    if (aAddress.IsLinkLocalAllNodesMulticast() || aAddress.IsRealmLocalAllNodesMulticast() ||
-        aAddress.IsRealmLocalAllMplForwarders())
+    for (NetifMulticastAddress *cur = mMulticastAddresses; cur; cur = cur->GetNext())
     {
-        ExitNow(rval = true);
+        if (memcmp(&cur->mAddress, &aAddress, sizeof(cur->mAddress)) == 0)
+        {
+            ExitNow(rval = true);
+        }
     }
-    else if (aAddress.IsLinkLocalAllRoutersMulticast() || aAddress.IsRealmLocalAllRoutersMulticast())
+
+exit:
+    return rval;
+}
+
+otError Netif::SubscribeAllRoutersMulticast(void)
+{
+    otError error = OT_ERROR_NONE;
+
+    if (mMulticastAddresses == &kLinkLocalAllNodesMulticastAddress)
     {
-        ExitNow(rval = mAllRoutersSubscribed);
+        mMulticastAddresses = static_cast<NetifMulticastAddress *>(
+                                  const_cast<otNetifMulticastAddress *>(&kRealmLocalAllRoutersMulticastAddress));
+    }
+    else
+    {
+        for (NetifMulticastAddress *cur = mMulticastAddresses; cur; cur = cur->GetNext())
+        {
+            if (cur == &kRealmLocalAllRoutersMulticastAddress)
+            {
+                ExitNow(error = OT_ERROR_ALREADY);
+            }
+
+            if (cur->mNext == &kLinkLocalAllNodesMulticastAddress)
+            {
+                cur->mNext = &kRealmLocalAllRoutersMulticastAddress;
```

Comment:
nit: but can we order the list so we're simply flipping between different group ids (e.g. `kRealmLocalAllNodesMulticastAddress` -> `kRealmLocalAllRoutersMulticastAddress`)?